runtime/gc_vlhgc/CopyForwardCompactGroup.hpp

```diff
@@ -126,12 +132,122 @@ class MM_CopyForwardCompactGroup
+		bool aligned = false;
+		if (_TLHRemainderBase) {
+			void *newAlignedTLHRemainderBase = (void*) MM_Math::roundToCeilingCard((uintptr_t) _TLHRemainderBase);
+			void *newAlignedTLHRemainderTop =  (void*) MM_Math::roundToFloorCard((uintptr_t) _TLHRemainderTop);
+			if (((uintptr_t)newAlignedTLHRemainderTop - (uintptr_t)newAlignedTLHRemainderBase) >= tlhSurvivorDiscardThreshold) {
+				/* TLHRemainder need to be card aligned for being reuse cross PGCs */
+				_TLHRemainderBase = newAlignedTLHRemainderBase;
+				_TLHRemainderTop = newAlignedTLHRemainderTop;
+				aligned = true;
+			}
+			/* else aligned TLHRemainder is smaller than tlhSurvivorDiscardThreshold, need to be reset */
+		}
+		return aligned;
+	}
+
+	MMINLINE bool abandonTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t compactGroup, bool preserveRemainder)
+	{
+		bool abandoned = true;
+		if (NULL != _TLHRemainderBase) {
+			Assert_MM_true(NULL != _TLHRemainderTop);
+			/* make it a walkable hole */
+			env->_cycleState->_activeSubSpace->abandonHeapChunk(_TLHRemainderBase, _TLHRemainderTop);
+			MM_GCExtensions* extensions =MM_GCExtensions::getExtensions(env);
+			MM_HeapRegionDescriptorVLHGC *region = (MM_HeapRegionDescriptorVLHGC*)extensions->heapRegionManager->tableDescriptorForAddress(_TLHRemainderBase);
+
+			/* only preserve for non Nursery regions */
+			if (!preserveRemainder || (MM_CompactGroupManager::getRegionAgeFromGroup(env, compactGroup) < extensions->tarokNurseryMaxAge._valueSpecified) || !region->containsObjects()) {
+				/* discard the remainder */
+				discardHeapChunk(_TLHRemainderBase, _TLHRemainderTop, region);
+				resetTLHRemainder();
+
+			} else {
+				_gcIDForTLHRemainder = extensions->globalVLHGCStats.gcCount;
+				abandoned = false;
+			}
+		} else {
+			Assert_MM_true(NULL == _TLHRemainderTop);
+		}
+		return abandoned;
+	}
+
+	MMINLINE void discardHeapChunk(void *base, void *top, MM_HeapRegionDescriptorVLHGC *region)
+	{
+		uintptr_t discardSize = ((uintptr_t)top - (uintptr_t)base);
+		_discardedBytes += discardSize;
+		region->getMemoryPool()->incrementDarkMatterBytes(discardSize);
+	}
+
+	/**
+	 * check and prepare TLHRemainder ready for using by collector
+	 *  1, reset TLHRemainder if it is invalid.
+	 *  2, reset TLHRemainder if it is in collection set
+	 *  3, align TLHRemainder with Card
+	 *  4, reset TLHRemainder if compactGroup doesn't match regionCompactGroup (the case: region overflow)
+     * @return region, if remainders are non-null, and regions is not already converted to survivor
+     * @return NULL, otherwise
+	 */
+	MMINLINE MM_HeapRegionDescriptorVLHGC* prepareTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t compactGroup)
+	{
+		MM_HeapRegionDescriptorVLHGC *region = NULL;
+		if (NULL != _TLHRemainderBase) {
+			bool needResetTLHRemainder = false;
+			MM_GCExtensions* extensions =MM_GCExtensions::getExtensions(env);
+			region = (MM_HeapRegionDescriptorVLHGC*)extensions->heapRegionManager->tableDescriptorForAddress(_TLHRemainderBase);
```

Comment:
we should assert that region's age matches compactGroup that is passed (the age extracted out of it)

Reply:
we can assert  region's age matches compactGroup, but we need to exclude the overflowed case, which the region age could jump directly to tenure (age24), currently we have the check for the case (line 237-239).

Reply:
ok, could we just put the assert in else case of if (regionCompactGroup != compactGroup)  ?